### PR TITLE
DOC: Document caveat in random.uniform

### LINKED ISSUE #10803: random.uniform documentation suggests `high` is excluded
There's a possibility of confusion from the `np.random.uniform` documentation, which says:

> Samples are uniformly distributed over the half-open interval [low, high) (includes low, but excludes high).

Arguably, that's true, in the sense that we're using an expression that (mathematically) draws from `[low, high)`, but rounding and numerical errors mean that in practice, `high` _can_ be generated:

```
>>> low = 1.0
>>> high = 1.0 + 2**-40
>>> low, high, low < high
(1.0, 1.0000000000009095, True)
>>> np.random.uniform(low, high, size=10**6).max() == high
True
```

The current documentation leads to people (well, at least one SO user) writing things like `np.random.uniform(low, np.nextafter(high, np.inf))` when they want a closed interval, which probably isn't a good idea:

```
>>> np.random.uniform(low, np.nextafter(high, np.inf), size=10**6).max() > high
True
```

### PR DESCRIPTION
**Pull request by kritisingh1**
Closes #10803

### CODE REVIEW

File: numpy/random/mtrand/mtrand.pyx [hunk 1]
```diff
@@ -1224,7 +1224,9 @@ cdef class RandomState:
         Draw samples from a uniform distribution.
 
         Samples are uniformly distributed over the half-open interval
-        ``[low, high)`` (includes low, but excludes high).  In other words,
+        ``[low, high)`` . The end point ``high`` may or may not be included
+        in the range depending on the floating-point rounding in the equation
+        ``(b-a) * random_sample() + a``.  In other words,
```
Comment (mattip) [inline numpy/random/mtrand/mtrand.pyx]: What are `a` and `b`? They are not arguments to `uniform`.

Comment (kritisingh1) [reply numpy/random/mtrand/mtrand.pyx]: Those were meant to be `low` and `high` respectively. I mistakenly took it directly from the `numpy.random.random_sample()` docs. Correcting.

**Comment by kritisingh1** [comment]
@mattip is there anything else I need to modify?

**Comment by mattip** [comment]
@kritisingh1 Perhaps the text can be improved. Under what conditions will `high` be included in the result? What is the range of the possible values of `random_sample` (hint: `dtype.eps`). Also, I am not sure the disclaimer is important enough to be in the first paragraph of the docstring, perhaps it should be a `.. note` after `Returns`. What do you think?

**Comment by kritisingh1** [comment]
Yes, I think a `.. note` after the `See Also` section would be the best place to add this since it already contains some information about the limits. I would look into how to make it more descriptive. Thanks!

**Comment by kritisingh1** [comment]
@mattip I've made a few changes. Please review. Thanks!

File: numpy/random/mtrand/mtrand.pyx [hunk 2]
```diff
@@ -1270,7 +1270,15 @@ cdef class RandomState:
...
-        inequality condition.
+        inequality condition. Also, if the ``size`` is large enough
+        ( 10^6 or more ), the ``high`` limit may be included in the
+        returned array of floats due to floating-point rounding in the
+        equation ``low + (high-low) * random_sample()``. For example:
+
+        >>> x = np.float32(5*0.99999999)
+        >>> x
+        5.0
+
```
Comment (mattip) [inline numpy/random/mtrand/mtrand.pyx]: It is not dependent on size, only on rounding. So something like "... inequality condition. The ``high`` limit may ...". The example is good.

**Comment by kritisingh1** [comment]
@mattip updated.

**Comment by kritisingh1** [comment]
@mattip is there anything else I need to change here?

**Comment by mattip** [comment]
@bashtage does this change look correct?

**Comment by bashtage** [comment]
Yes, correct.  Not sure if it is worth discussing a mitigation strategy.

**Comment by mdickinson** [comment]
It looks as though this change is no longer present in `mtrand.pyx` in master. Any idea what happened?

**Comment by bashtage** [comment]
This was made in the 1.16 mtrand.pyx file which was replaced by _mtrand.pyx for 1.17 that shares some of the new Generator code.  This doc change got lost in the move it seems.

**Comment by bashtage** [comment]
Looking at blame, it seems it got lost in the move over from the old mtrand to the new implementation.

**Comment by mattip** [comment]
I wonder what else got lost.

In changeset c53b2eb72 this file was removed in favour of the randomgen version, without merging. I think only `mtrand.pyx` has user-facing code from that changeset, so I will compare the two files and try to forward port this change and any others that we missed.

**Comment by mattip** [comment]
xref gh-15575

**Pull request merged**